sonarts-core/src/rules/noGratuitousExpressionsRule.ts

```diff
@@ -0,0 +1,122 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import * as tslint from "tslint";
+import * as ts from "typescript";
+import * as tsutils from "tsutils";
+import { SonarRuleMetaData } from "../sonarRule";
+import { execute } from "../se/SymbolicExecution";
+import { build } from "../cfg/builder";
+import { ProgramState, createInitialState } from "../se/programStates";
+import { isTruthy, Constraint, isFalsy } from "../se/constraints";
+import { SymbolTableBuilder } from "../symbols/builder";
+import { SymbolTable, UsageFlag } from "../symbols/table";
+import { firstLocalAncestor, FUNCTION_LIKE, is } from "../utils/navigation";
+
+export class Rule extends tslint.Rules.TypedRule {
+  public static metadata: SonarRuleMetaData = {
+    ruleName: "no-gratuitous-expressions",
+    description: "Boolean expressions should not be gratuitous",
+    rationale: tslint.Utils.dedent`
+      If a boolean expression doesn't change the evaluation of the condition, then it is entirely unnecessary, and can 
+      be removed. If it is gratuitous because it does not match the programmer's intent, then it's a bug and 
+      the expression should be fixed.`,
+    optionsDescription: "",
+    options: null,
+    rspecKey: "RSPEC-2589",
+    type: "functionality",
+    typescriptOnly: false,
+  };
+
+  public static getMessage(value: string) {
+    return `This condition always evaluates to "${value}".`;
+  }
+
+  public applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): tslint.RuleFailure[] {
+    const symbols = SymbolTableBuilder.build(sourceFile, program);
+    return this.applyWithWalker(new Walker(sourceFile, this.getOptions(), program, symbols));
+  }
+}
+
+class Walker extends tslint.ProgramAwareRuleWalker {
+  public constructor(
+    sourceFile: ts.SourceFile,
+    options: tslint.IOptions,
+    program: ts.Program,
+    private readonly symbols: SymbolTable,
+  ) {
+    super(sourceFile, options, program);
+  }
+
+  protected visitNode(node: ts.Node) {
+    if (is(node, ...FUNCTION_LIKE)) {
+      const functionLike = node as ts.FunctionLikeDeclaration;
+      const statements = this.getStatements(functionLike);
+      const initialState = createInitialState(functionLike, this.getProgram());
+      const shouldTrackSymbol = (symbol: ts.Symbol) =>
```

Comment:
isn't it not optimal to have a function instead of set?

Reply:
@vilchik-elena maybe, but this way it covers elegantly the case where we simply do not care to check is we should track a symbol (unit tests and SourceFile SE). Also, it should either disappear after we introduce LVA or move out of the rule and into SE when we have more than one rule using SE. I would keep it as it is and wait for one of those two events to change it.